File: golang/ast_extract.go
// Deterministic Go AST facts for todo2code. Uses only the Go standard library.
//
// Emits the same JSON envelope as python/ast_extract.py so the TypeScript
// extractor can treat every language adapter identically:
//
//	{"facts": [...], "warnings": [...]}
//
// Run standalone (no module required):
//
//	go run golang/ast_extract.go <root> --max-file-bytes 524288
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

var ignoredDirs = map[string]bool{
	".git": true, "node_modules": true, "dist": true, "build": true,
	"coverage": true, ".intent": true, "vendor": true, "testdata": true,
}

// Fact mirrors the PythonFact interface consumed by src/extractors/ast.ts.
type Fact struct {
	Path        string         `json:"path"`
	LineStart   int            `json:"lineStart"`
	LineEnd     int            `json:"lineEnd"`
	Kind        string         `json:"kind"`
	Action      string         `json:"action"`
	Object      string         `json:"object"`
	Symbol      *string        `json:"symbol"`
	Subject     *string        `json:"subject"`
	Excerpt     string         `json:"excerpt"`
	ContentHash string         `json:"contentHash"`
	Metadata    map[string]any `json:"metadata"`
}

type output struct {
	Facts    []Fact   `json:"facts"`
	Warnings []string `json:"warnings"`
}

func main() {
	root := "."
	maxFileBytes := int64(524288)

	args := os.Args[1:]
	for index := 0; index < len(args); index++ {
		switch args[index] {
		case "--max-file-bytes":
			if index+1 < len(args) {
				if parsed, err := strconv.ParseInt(args[index+1], 10, 64); err == nil {
					maxFileBytes = parsed
				}
				index++
			}
		default:
			if !strings.HasPrefix(args[index], "-") {
				root = args[index]
			}
		}
	}

	result := output{Facts: []Fact{}, Warnings: []string{}}
	files, err := collectGoFiles(root)
	if err != nil {
		result.Warnings = append(result.Warnings, fmt.Sprintf("walk failed: %v", err))
		emit(result)
		return
	}

	for _, file := range files {
		info, err := os.Stat(file)
		if err != nil {
			result.Warnings = append(result.Warnings, fmt.Sprintf("%s: %v", file, err))
			continue
		}
		if info.Size() > maxFileBytes {
			relative, _ := filepath.Rel(root, file)
			result.Warnings = append(result.Warnings,
				fmt.Sprintf("%s: skipped, %d bytes exceeds limit %d", toSlash(relative), info.Size(), maxFileBytes))
			continue
		}

		source, err := os.ReadFile(file)
		if err != nil {
			result.Warnings = append(result.Warnings, fmt.Sprintf("%s: %v", file, err))
			continue
		}
		relative, err := filepath.Rel(root, file)
		if err != nil {
			relative = file
		}
		facts, warning := parseFile(toSlash(relative), source)
		if warning != "" {
			result.Warnings = append(result.Warnings, warning)
			continue
		}
		result.Facts = append(result.Facts, facts...)
	}

	emit(result)
}

func emit(result output) {
	encoder := json.NewEncoder(os.Stdout)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(result); err != nil {
		fmt.Fprintf(os.Stderr, "encode failed: %v\n", err)
		os.Exit(1)
	}
}

func collectGoFiles(root string) ([]string, error) {
	var files []string
	err := filepath.Walk(root, func(current string, info os.FileInfo, err error) error {
		if err != nil {
			return nil // Skip unreadable entries rather than aborting the walk.
		}
		if info.IsDir() {
			if ignoredDirs[info.Name()] || (strings.HasPrefix(info.Name(), ".") && info.Name() != "." && current != root) {
				return filepath.SkipDir
			}
			return nil
		}
		if strings.HasSuffix(info.Name(), ".go") {
			files = append(files, current)
		}
		return nil
	})
	sort.Strings(files)
	return files, err
}

func parseFile(relative string, source []byte) ([]Fact, string) {
	fileSet := token.NewFileSet()
	parsed, err := parser.ParseFile(fileSet, relative, source, parser.ParseComments)
	if err != nil {
		return nil, fmt.Sprintf("%s: parse error: %v", relative, err)
	}

	lines := strings.Split(string(source), "\n")
	collector := &factCollector{relative: relative, fileSet: fileSet, lines: lines}

	collector.add(parsed.Name, "go_package_fact", "declare", parsed.Name.Name,
		strPtr(parsed.Name.Name), nil, map[string]any{"package": parsed.Name.Name})

	for _, importSpec := range parsed.Imports {
		importPath, err := strconv.Unquote(importSpec.Path.Value)
		if err != nil {
			importPath = importSpec.Path.Value
		}
		metadata := map[string]any{"package": parsed.Name.Name}
		if importSpec.Name != nil {
			metadata["alias"] = importSpec.Name.Name
		}
		collector.add(importSpec, "go_import_fact", "depend_on", importPath, nil, nil, metadata)
	}

	for _, declaration := range parsed.Decls {
		collector.visitDecl(declaration, parsed.Name.Name)
	}

	return collector.facts, ""
}

type factCollector struct {
	relative string
	fileSet  *token.FileSet
	lines    []string
	facts    []Fact
}

func (c *factCollector) position(node ast.Node) (int, int) {
	start := c.fileSet.Position(node.Pos()).Line
	end := c.fileSet.Position(node.End()).Line
	if start < 1 {
		start = 1
	}
	if end < start {
		end = start
	}
	return start, end
}

func (c *factCollector) excerpt(start, end int) string {
	if start > len(c.lines) {
		return ""
	}
	if end > len(c.lines) {
		end = len(c.lines)
	}
	text := strings.Join(c.lines[start-1:end], "\n")
	if len(text) > 2000 {
		text = text[:2000]
	}
	return text
}

func (c *factCollector) add(node ast.Node, kind, action, object string, symbol, subject *string, metadata map[string]any) {
	start, end := c.position(node)
	excerpt := c.excerpt(start, end)
	sum := sha256.Sum256([]byte(excerpt))
	if metadata == nil {
		metadata = map[string]any{}
	}
	c.facts = append(c.facts, Fact{
		Path:        c.relative,
		LineStart:   start,
		LineEnd:     end,
		Kind:        kind,
		Action:      action,
		Object:      object,
		Symbol:      symbol,
		Subject:     subject,
		Excerpt:     excerpt,
		ContentHash: hex.EncodeToString(sum[:]),
		Metadata:    metadata,
	})
}

func (c *factCollector) visitDecl(declaration ast.Decl, packageName string) {
	switch node := declaration.(type) {
	case *ast.FuncDecl:
		c.visitFunc(node, packageName)
	case *ast.GenDecl:
		c.visitGenDecl(node, packageName)
	}
}

func (c *factCollector) visitFunc(node *ast.FuncDecl, packageName string) {
	name := node.Name.Name
	metadata := map[string]any{
		"package":  packageName,
		"exported": node.Name.IsExported(),
		"kind":     "function",
	}

	// A method carries its receiver type, which is what makes `Store.Enqueue`
	// resolvable from a TODO or commit message.
	var subject *string
	if node.Recv != nil && len(node.Recv.List) > 0 {
		receiver := typeName(node.Recv.List[0].Type)
		if receiver != "" {
			metadata["kind"] = "method"
			metadata["receiver"] = receiver
			name = receiver + "." + name
			subject = strPtr(receiver)
		}
	}
	if node.Type.Params != nil {
		metadata["parameters"] = len(node.Type.Params.List)
	}

	c.add(node, "go_symbol_fact", "declare", name, strPtr(name), subject, metadata)

	if node.Body != nil {
		c.visitCalls(node.Body, name, packageName)
	}
}

func (c *factCollector) visitGenDecl(node *ast.GenDecl, packageName string) {
	for _, spec := range node.Specs {
		switch value := spec.(type) {
		case *ast.TypeSpec:
			metadata := map[string]any{
				"package":  packageName,
				"exported": value.Name.IsExported(),
				"kind":     declaredTypeKind(value.Type),
			}
			c.add(value, "go_symbol_fact", "declare", value.Name.Name, strPtr(value.Name.Name), nil, metadata)
		case *ast.ValueSpec:
			for _, name := range value.Names {
				if name.Name == "_" {
					continue
				}
				kind := "var"
				if node.Tok == token.CONST {
					kind = "const"
				}
				metadata := map[string]any{
					"package":  packageName,
					"exported": name.IsExported(),
					"kind":     kind,
				}
				c.add(name, "go_symbol_fact", "declare", name.Name, strPtr(name.Name), nil, metadata)
			}
		}
	}
}

func (c *factCollector) visitCalls(body *ast.BlockStmt, scope, packageName string) {
	ast.Inspect(body, func(node ast.Node) bool {
		call, ok := node.(*ast.CallExpr)
		if !ok {
			return true
		}
		callee := typeName(call.Fun)
		if callee == "" {
			return true
		}
		c.add(call, "go_call_fact", "call", callee, strPtr(scope), nil, map[string]any{
			"callee":  callee,
			"package": packageName,
		})
		return true
	})
}

// typeName renders the dotted name of an expression, which covers identifiers,
// selectors, pointers, generics and the common composite forms.
func typeName(expr ast.Expr) string {
	switch node := expr.(type) {
	case *ast.Ident:
		return node.Name
	case *ast.SelectorExpr:
		left := typeName(node.X)
		if left == "" {
			return node.Sel.Name
		}
		return left + "." + node.Sel.Name
	case *ast.StarExpr:
		return typeName(node.X)
	case *ast.IndexExpr:
		return typeName(node.X)
	case *ast.IndexListExpr:
		return typeName(node.X)
	case *ast.ParenExpr:
		return typeName(node.X)
	case *ast.ArrayType:
		return typeName(node.Elt)
	case *ast.Ellipsis:
		return typeName(node.Elt)
	default:
		return ""
	}
}

func declaredTypeKind(expr ast.Expr) string {
	switch expr.(type) {
	case *ast.StructType:
		return "struct"
	case *ast.InterfaceType:
		return "interface"
	case *ast.FuncType:
		return "func_type"
	default:
		return "type"
	}
}

func strPtr(value string) *string {
	return &value
}

func toSlash(value string) string {
	return filepath.ToSlash(value)
}

File: python/ast_extract.py
#!/usr/bin/env python3
"""Deterministic Python AST facts for todo2code. Uses only the Python standard library."""
from __future__ import annotations

import argparse
import ast
import hashlib
import json
import os
import sys
from pathlib import Path
from typing import Any

IGNORED_DIRS = {'.git', 'node_modules', 'dist', 'build', 'coverage', '.intent', '.venv', 'venv', '__pycache__'}


def source_hash(value: str) -> str:
    return hashlib.sha256(value.encode('utf-8')).hexdigest()


def dotted_name(node: ast.AST) -> str:
    if isinstance(node, ast.Name):
        return node.id
    if isinstance(node, ast.Attribute):
        left = dotted_name(node.value)
        return f'{left}.{node.attr}' if left else node.attr
    if isinstance(node, ast.Call):
        return dotted_name(node.func)
    return ''


def is_module_entrypoint(node: ast.If) -> bool:
    """Return true for the canonical ``if __name__ == '__main__'`` guard."""
    test = node.test
    if not isinstance(test, ast.Compare) or len(test.ops) != 1 or len(test.comparators) != 1:
        return False
    if not isinstance(test.ops[0], ast.Eq):
        return False
    values = (test.left, test.comparators[0])
    has_name = any(isinstance(value, ast.Name) and value.id == '__name__' for value in values)
    has_main = any(isinstance(value, ast.Constant) and value.value == '__main__' for value in values)
    return has_name and has_main


class FactVisitor(ast.NodeVisitor):
    def __init__(self, relative_path: str, lines: list[str]) -> None:
        self.relative_path = relative_path
        self.lines = lines
        self.scope: list[str] = []
        self.function_depth = 0
        self.facts: list[dict[str, Any]] = []

    def excerpt(self, node: ast.AST) -> str:
        start = max(1, getattr(node, 'lineno', 1))
        end = max(start, getattr(node, 'end_lineno', start))
        return '\n'.join(self.lines[start - 1:min(end, len(self.lines))])[:2000]

    def add(self, node: ast.AST, kind: str, action: str, obj: str, symbol: str | None = None, metadata: dict[str, Any] | None = None) -> None:
        start = max(1, getattr(node, 'lineno', 1))
        end = max(start, getattr(node, 'end_lineno', start))
        excerpt = self.excerpt(node)
        self.facts.append({
            'path': self.relative_path,
            'lineStart': start,
            'lineEnd': end,
            'kind': kind,
            'action': action,
            'object': obj or 'unknown',
            'symbol': symbol,
            'subject': '.'.join(self.scope) if self.scope else None,
            'excerpt': excerpt,
            'contentHash': source_hash(excerpt),
            'metadata': metadata or {},
        })

    def visit_Import(self, node: ast.Import) -> Any:
        for alias in node.names:
            self.add(node, 'python_import_fact', 'depend_on', alias.name, metadata={'alias': alias.asname})
        self.generic_visit(node)

    def visit_ImportFrom(self, node: ast.ImportFrom) -> Any:
        module = node.module or ''
        for alias in node.names:
            name = f'{module}.{alias.name}'.strip('.')
            self.add(node, 'python_import_fact', 'depend_on', name, metadata={'alias': alias.asname, 'level': node.level})
        self.generic_visit(node)

    def visit_FunctionDef(self, node: ast.FunctionDef) -> Any:
        self.add(node, 'python_symbol_fact', 'declare', node.name, symbol=node.name, metadata={
            'symbolKind': 'function',
            'async': False,
            'decorators': [dotted_name(item) for item in node.decorator_list],
            'arguments': [item.arg for item in node.args.args],
        })
        self.scope.append(node.name)
        self.function_depth += 1
        self.generic_visit(node)
        self.function_depth -= 1
        self.scope.pop()

    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> Any:
        self.add(node, 'python_symbol_fact', 'declare', node.name, symbol=node.name, metadata={
            'symbolKind': 'function',
            'async': True,
            'decorators': [dotted_name(item) for item in node.decorator_list],
            'arguments': [item.arg for item in node.args.args],
        })
        self.scope.append(node.name)
        self.function_depth += 1
        self.generic_visit(node)
        self.function_depth -= 1
        self.scope.pop()

    def visit_ClassDef(self, node: ast.ClassDef) -> Any:
        self.add(node, 'python_symbol_fact', 'declare', node.name, symbol=node.name, metadata={
            'symbolKind': 'class',
            'bases': [dotted_name(item) for item in node.bases],
            'decorators': [dotted_name(item) for item in node.decorator_list],
        })
        self.scope.append(node.name)
        self.generic_visit(node)
        self.scope.pop()

    def add_named_constant(self, node: ast.AST, name: str, value: ast.AST) -> None:
        """Record module/class constants, but never ordinary local variables."""
        if self.function_depth > 0 or not name.isupper():
            return
        try:
            literal = ast.literal_eval(value)
        except (ValueError, TypeError, MemoryError, RecursionError):
            return
        if literal is None or isinstance(literal, bool) or not isinstance(literal, (int, float, str)):
            return
        rendered = repr(literal)
        if len(rendered) > 120:
            return
        self.add(node, 'python_constant_fact', 'declare', f'named constant {rendered}', symbol=name, metadata={
            'symbolKind': 'constant',
            'constantName': name,
            'constantValue': literal,
        })

    def visit_Assign(self, node: ast.Assign) -> Any:
        for target in node.targets:
            if isinstance(target, ast.Name):
                self.add_named_constant(node, target.id, node.value)
        self.generic_visit(node)

    def visit_AnnAssign(self, node: ast.AnnAssign) -> Any:
        if isinstance(node.target, ast.Name) and node.value is not None:
            self.add_named_constant(node, node.target.id, node.value)
        self.generic_visit(node)

    def visit_If(self, node: ast.If) -> Any:
        if self.function_depth == 0 and is_module_entrypoint(node):
            self.add(node, 'python_module_entrypoint_fact', 'declare', 'module execution', metadata={
                'symbolKind': 'module_entrypoint',
            })
        self.generic_visit(node)

    def visit_Call(self, node: ast.Call) -> Any:
        callee = dotted_name(node.func)
        if callee:
            self.add(node, 'python_call_fact', 'call', callee, symbol='.'.join(self.scope) or None, metadata={'callee': callee})
        self.generic_visit(node)


def iter_python_files(root: Path, files_from: Path | None = None) -> list[Path]:
    if files_from is not None:
    return _iter_python_files_from_manifest(root, files_from) if files_from is not None else _iter_python_files_from_disk(root)


def _iter_python_files_from_manifest(root: Path, files_from: Path) -> list[Path]:
    values = json.loads(files_from.read_text(encoding='utf-8'))
    if not isinstance(values, list) or any(not isinstance(value, str) for value in values):
        raise ValueError('--files-from must contain a JSON array of paths')
    output: list[Path] = []
    for value in values:
        candidate = _resolve_candidate(root, value)
        if candidate.suffix == '.py' and candidate.is_file() and not candidate.is_symlink():
            output.append(candidate)
    return sorted(set(output))


def _iter_python_files_from_disk(root: Path) -> list[Path]:
    output: list[Path] = []
    for current, directories, files in os.walk(root):
        directories[:] = sorted(directory for directory in directories if directory not in IGNORED_DIRS)
        for filename in sorted(files):
            if filename.endswith('.py'):
                candidate = Path(current) / filename
                if not candidate.is_symlink():
                    output.append(candidate)
    return output


def _resolve_candidate(root: Path, value: str) -> Path:
    candidate = (root / value).resolve()
    try:
        candidate.relative_to(root)
    except ValueError as exc:
        raise ValueError(f'Python source escapes root: {value}') from exc
    return candidate


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('root')
    parser.add_argument('--max-file-bytes', type=int, default=524288)
    parser.add_argument('--files-from', type=Path)
    args = parser.parse_args()
    root = Path(args.root).resolve()
    facts: list[dict[str, Any]] = []
    warnings: list[str] = []
    for file_path in iter_python_files(root, args.files_from):
        try:
            if file_path.stat().st_size > args.max_file_bytes:
                warnings.append(f'skipped oversized Python file: {file_path.relative_to(root)}')
                continue
            body = file_path.read_text(encoding='utf-8')
            tree = ast.parse(body, filename=str(file_path), type_comments=True)
            visitor = FactVisitor(file_path.relative_to(root).as_posix(), body.splitlines())
            visitor.visit(tree)
            facts.extend(visitor.facts)
        except (SyntaxError, UnicodeDecodeError, OSError) as exc:
            warnings.append(f'{file_path.relative_to(root)}: {exc}')
    print(json.dumps({'facts': facts, 'warnings': warnings}, ensure_ascii=False, sort_keys=True))
    return 0


if __name__ == '__main__':
    sys.exit(main())
